mesonbuild/msubprojects.py

```diff
@@ -47,6 +47,9 @@ def git_output(cmd, workingdir):
     return quiet_git(cmd, workingdir, check=True)[1]
 
 def git_stash(workingdir):
+    out = git_output(['diff', '-s'], workingdir)
+    if not out.strip():
+        return
```

Comment:
This will discard changes that have been staged but not committed.

Reply:
Actually this will _always_ discard output. `-s` is the opposite of `-p`, and by default it will cause `git diff` to output nothing. Probably want to add a test for this now that we know this can break? :)

Reply:
Ooops, that's a typo because I initially wrote `git status -s`. Changed to --stat. Good catch!

Yes we definitely need tests, but it's a bit complicated to simulate those git repositories. Do you know if we have helpers somewhere already? Sadly git wraps is largely untested atm I think.

Reply:
No, this change is not enough. It will discard changes that have been staged. Observe:

```
$ vim meson.py
$ git diff --stat
 meson.py | 1 +
 1 file changed, 1 insertion(+)
$ git add meson.py
$ git diff --stat
$
```

Reply:
Also `git status -s` will also show untracked files, which we *do* want to ignore.

Reply:
`git status --short --untracked-files=no --ignore-submodules=untracked` looks like the best option, I think.

Reply:
`git diff --quiet HEAD`

By default git-diff(1) shows the difference between the working tree and the staging index. Adding --stat merely suppresses the line diff, which you promptly ignore for performance reasons I assume. It still returns "some" stdout in and *only* in situations where git-diff(1) would.

- `--quiet` tells git-diff(1) to completely avoid emitting output. Instead, it emits a useful return code for scripting.
- Passing HEAD as the diff target will make the comparison be between the working tree and the HEAD commit; it therefore shows both staged and unstaged changes.

Reply:
> No, this change is not enough. It will discard changes that have been staged. Observe:
Good point.

> Also `git status -s` will also show untracked files, which we _do_ want to ignore:
Right, that's why I changed to git diff.

@eli-schwartz thanks for the tip, I think `git diff --quiet HEAD` would be best then, I'll update.

Reply:
Fixed, and covered by unit test.